xrootd/protocol/protocol/protocol.go

```diff
@@ -7,134 +7,234 @@
 )
 
-// SecurityOptions are the security-related options.
+// securityOptions are the security-related options.
 // See specification for details: http://xrootd.org/doc/dev45/XRdv310.pdf, p. 72.
-type SecurityOptions byte
+type securityOptions byte
 
 const (
-	// None specifies that no security options are provided.
-	None SecurityOptions = 0
-	// ForceSecurity specifies that signing is required even if the authentication
+	// forceSecurity specifies that signing is required even if the authentication
 	// protocol does not support generic encryption.
-	ForceSecurity SecurityOptions = 0x02
+	forceSecurity securityOptions = 0x02
 )
 
-// SecurityInfoLength is the length of SecurityInfo in bytes.
-const SecurityInfoLength = 6
+// requestOptions specifies what should be returned as part of response.
+type requestOptions byte
 
-// SecurityInfo is a response part that is provided when required (if server supports that).
-// It contains the security version, the security options, the security level,
-// and the number of following security overrides, if any.
-type SecurityInfo struct {
-	// FIXME: Rename Reserved* fields to _ when automatically generated (un)marshalling will be available.
-	Reserved1             byte
-	Reserved2             byte
+const (
+	// requestOptionsNone specifies that only general response should be returned.
+	requestOptionsNone requestOptions = 0
+	// returnSecurityRequirements specifies that security requirements should be returned
+	// if that's supported by the server.
+	returnSecurityRequirements requestOptions = 1
+)
+
+// Request holds protocol request parameters.
+type Request struct {
+	ClientProtocolVersion int32
+	Options               requestOptions
+	_                     [11]byte
+	_                     int32
+}
+
+// NewRequest forms a Request according to provided parameters.
+func NewRequest(protocolVersion int32, withSecurityRequirements bool) *Request {
+	var options = requestOptionsNone
+	if withSecurityRequirements {
+		options |= returnSecurityRequirements
+	}
+	return &Request{ClientProtocolVersion: protocolVersion, Options: options}
+}
+
+// ReqID implements protocol.Request.ReqID
+func (req *Request) ReqID() uint16 { return RequestID }
+
+// MarshalXrd implements xrootd/protocol.Marshaler
+func (o Request) MarshalXrd() (data []byte, err error) {
+	var enc xrdenc.Encoder
+	enc.WriteI32(o.ClientProtocolVersion)
+	enc.WriteU8(byte(o.Options))
+	enc.WriteReserved(11)
+	enc.WriteReserved(4)
+	return enc.Bytes(), err
+}
+
+// UnmarshalXrd implements xrootd/protocol.Unmarshaler
+func (o *Request) UnmarshalXrd(data []byte) (pos int, err error) {
+	dec := xrdenc.NewDecoder(data)
+	o.ClientProtocolVersion = dec.ReadI32()
+	o.Options = requestOptions(dec.ReadU8())
+	dec.Skip(11)
+	dec.Skip(4)
+	return dec.Pos(), err
+}
+
+// Response is a response for the `Protocol` request. See details in the xrootd protocol specification.
+type Response struct {
```

Comment:
somehow, the `SetIsXyz` methods on this `protocol.Response` type feel a bit wrong to me: calling those won't impact back the server from which we got them.
I understand you introduced them so you could create a `Response` to test against in the `Client` tests.

Perhaps we could just leave the `flags` field of `Response` exported (so `s/flags/Flags/`) and re-export the `IsXyz` bitmasks ?
the `protocol` package isn't super crowded so it's fine to populate it somewhat.

that way, you could create a `want` value like so:
```go
want := protocol.Response{
￼	BinaryProtocolVersion: protocolVersion,
￼	Flags:                 protocol.IsManager | protocol.IsServer | protocol.IsMeta |
                               protocol.IsProxy   | protocol.IsSupervisor,
	SecurityLevel:         xrdproto.Pedantic,
	SecurityOverrides:     []xrdproto.SecurityOverride{{1, xrdproto.SignNeeded}},
}
```

Reply:
>I understand you introduced them so you could create a Response to test against in the Client tests.

was thinking about using them for the server implementation, not just tests. However, I don't argue about usage of `Flags`, it's fine too. :)